### PR TITLE
[Meson] Build without werror by default

### PR DESCRIPTION
**Pull request by mkow**
## Description of the changes 

Compiler warnings are not stable and sometimes based on heuristics, so we shouldn't fail the build on warnings by default (this can easily cause Gramine to be unbuildable on new compilers/distros due to trivial or non-issues, see e.g. https://github.com/gramineproject/gramine/pull/220).

## How to test this PR? 

Build with and without passing `--werror` to `meson setup`.

---
This change is [Reviewable](https://reviewable.io/reviews/gramineproject/gramine/409)

### CODE REVIEW

**Comment by eli-schwartz** [comment]
You may want to pass that option explicitly in CI instead. :)

... also you don't need to pass `werror=false` in default options, simply delete the line altogether as that is the default anyway.

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/409)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @mkow)

___
*[meson.build, line 16 at r1](https://reviewable.io/reviews/gramineproject/gramine/409#-Mw0aS-M0VMM6kvp4Srg:-Mw0aS-NB1K9UXfes4XL:bq24a67) ([raw file](https://github.com/gramineproject/gramine/blob/6c82dc4a6063857d7f94b4cfe95c5b34d3fcf77f/meson.build#L16)):*
> ```Meson
>         'c_std=c11',
>         'cpp_std=c++14',
>         'werror=false',
> ```

But why? We'll end up with PRs containing legit issues, that could have been easily caught by `werror=true`.

I see obvious downsides to switching the current logic off, and I don't see obvious upsides... Forcing people to double-check their code on warnings sounds good to me.

**Review by woju — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/409)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @dimakuv)

___
*[meson.build, line 16 at r1](https://reviewable.io/reviews/gramineproject/gramine/409#-Mw0aS-M0VMM6kvp4Srg:-Mw1P-yVA2NNTEK_DQvD:bmfmktq) ([raw file](https://github.com/gramineproject/gramine/blob/6c82dc4a6063857d7f94b4cfe95c5b34d3fcf77f/meson.build#L16)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

But why? We'll end up with PRs containing legit issues, that could have been easily caught by `werror=true`.

I see obvious downsides to switching the current logic off, and I don't see obvious upsides... Forcing people to double-check their code on warnings sounds good to me.
</blockquote></details>

Seconded. I think this is overreaction.

**Review by mkow — COMMENTED** [review]
> You may want to pass that option explicitly in CI instead. :)

We do pass it in CI, although using envs: https://github.com/gramineproject/gramine/blob/d2cb06f6c69ad32d669acf10abbce15cd6d4c0b1/.ci/lib/config.jenkinsfile#L1
Not sure why we're doing it this way, I guess @woju will know :)
Or maybe I misunderstood what you mean?

> ... also you don't need to pass `werror=false` in default options, simply delete the line altogether as that is the default anyway.

Yeah, I was thinking about two sides of the spectrum: skipping it altogether, or keeping it and even adding an explicit comment that this setting is intentional. Chose something in the middle in the end, but I'm not attached to this choice.
@dimakuv, @boryspoplawski, @pwmarcz, @woju: any preferences?

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/409#-:-Mw1nbhZ-h5YpSPw9sX4:b-p9o3gr)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @dimakuv and @woju)

___
*[meson.build, line 16 at r1](https://reviewable.io/reviews/gramineproject/gramine/409#-Mw0aS-M0VMM6kvp4Srg:-Mw1ohs0EQX1Zj1IA6Rb:b6tstk4) ([raw file](https://github.com/gramineproject/gramine/blob/6c82dc4a6063857d7f94b4cfe95c5b34d3fcf77f/meson.build#L16)):*
<details><summary><i>Previously, woju (Wojtek Porczyk) wrote…</i></summary><blockquote>

Seconded. I think this is overreaction.
</blockquote></details>

The reason is that we often want people to use / test the current master (e.g. when reporting issues). And if we ask people to do this, then we should ensure that it's stable. And werror is often stable nor reasonable.

If we want contributors to use it, then maybe just say this in CONTRIBUTING.rst? I just want master branch to be stable and functional for everyone. Cloning an open source project and building it only to see the build fail despite doing everything correctly is a really shitty experience and doesn't leave a good impression on the project.

Some twitter thread I fully agree with: https://twitter.com/RichFelker/status/1435374208155017218.

**Review by pwmarcz — COMMENTED** [review]
`env.WERROR` sounds like a leftover from Makefile-based build system. Is it even connected to anything? `git grep WERROR` gives me no other hits.

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/409#-:-Mw1rd59D3RgZczSwTjk:b6bzbdl)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @dimakuv and @woju)

___
*[meson.build, line 16 at r1](https://reviewable.io/reviews/gramineproject/gramine/409#-Mw0aS-M0VMM6kvp4Srg:-Mw1rnVa08Q8nkqI2P4S:b1rpwh6) ([raw file](https://github.com/gramineproject/gramine/blob/6c82dc4a6063857d7f94b4cfe95c5b34d3fcf77f/meson.build#L16)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

The reason is that we often want people to use / test the current master (e.g. when reporting issues). And if we ask people to do this, then we should ensure that it's stable. And werror is often stable nor reasonable.

If we want contributors to use it, then maybe just say this in CONTRIBUTING.rst? I just want master branch to be stable and functional for everyone. Cloning an open source project and building it only to see the build fail despite doing everything correctly is a really shitty experience and doesn't leave a good impression on the project.

Some twitter thread I fully agree with: https://twitter.com/RichFelker/status/1435374208155017218.
</blockquote></details>

I agree with @mkow that distributing source with `-Werror` disabled is a sensible default. But if we do that, we should make sure that:

1. It's enabled in CI (as said above, I don't think it is in this PR).
2. It's enabled by default in whatever package build process we have. If we build a package, and there are warnings, we don't want to ignore them.
3. As a maintainer or contributor, it's easy for me to have it always enabled.

We could use `meson setup --werror` in all our CI scripts and package build scripts, mention `--werror` in the build documentation, and say in CONTRIBUTING.rst that we require it if you contribute code.

**Review by woju — CHANGES_REQUESTED** [review]
> We do pass it in CI, although using envs

Yes, this is a remnant from older buildsystem, we didn't clean up after Makefiles.

The problem with our CI is, most contributors don't get to see the error messages, because Jenkins is firewalled (for one too many security issue). Reviewers would have to manually paste the relevant errors to PR authors. We plan to rework the CI, but until then, I think it would be easier to just force `werror=true` down people's throats. Unfortunately.

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/409#-:-Mw1uYJS6Z058s6okQR-:bh1w9mn)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @dimakuv)

___
*[meson.build, line 16 at r1](https://reviewable.io/reviews/gramineproject/gramine/409#-Mw0aS-M0VMM6kvp4Srg:-Mw1xICh9xmm5Rde2Lsn:b-ff1r57) ([raw file](https://github.com/gramineproject/gramine/blob/6c82dc4a6063857d7f94b4cfe95c5b34d3fcf77f/meson.build#L16)):*
<details><summary><i>Previously, pwmarcz (Paweł Marczewski) wrote…</i></summary><blockquote>

I agree with @mkow that distributing source with `-Werror` disabled is a sensible default. But if we do that, we should make sure that:

1. It's enabled in CI (as said above, I don't think it is in this PR).
2. It's enabled by default in whatever package build process we have. If we build a package, and there are warnings, we don't want to ignore them.
3. As a maintainer or contributor, it's easy for me to have it always enabled.

We could use `meson setup --werror` in all our CI scripts and package build scripts, mention `--werror` in the build documentation, and say in CONTRIBUTING.rst that we require it if you contribute code.
</blockquote></details>

I'm not yet convinced. Do you think people will really use it if you just write about it in README?

And yes, if we want werror to be disabled by default, please at least re-enable it in `stage-build*` by adding `--werror` to `meson setup`. (And clean up `env.WERROR` :) ).

**Review by mkow — COMMENTED** [review]
> Nah, I just rolled a Nat 1 at reading your CI configs. :D (Edit: or maybe not? Don't ask me...)

Haha :) But @pwmarcz may be right, I now got more suspicious at this env.

> Reviewers would have to manually paste the relevant errors to PR authors.

We do this anyways very often because people usually don't run all the tests themselves ;) So I don't think this is a big drawback. And we can always ask in CONTRIBUTING.rst to use `--werror`, same as we ask many other things.

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/409#-:-Mw1wiC592nxtAimBd3u:bf5bow0)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @mkow)

___
*[meson.build, line 16 at r1](https://reviewable.io/reviews/gramineproject/gramine/409#-Mw0aS-M0VMM6kvp4Srg:-Mw1ya3F9uRGZ1Vav3cO:b-4bddra) ([raw file](https://github.com/gramineproject/gramine/blob/6c82dc4a6063857d7f94b4cfe95c5b34d3fcf77f/meson.build#L16)):*
Yeah, I'll fix this `env.WERROR`. Looks like Make remnants.

> Do you think people will really use it if you just write about it in README?

The same goes for everything else, like running tests before submitting a PR. We ask for this and then enforce in CI. People often don't do this, but then we can easily see in CI that they didn't. And IMO working by-default is more important than having this small improvement for people who don't read CONTRIBUTING.rst.

**Review by mkow — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/409)* status: 1 of 7 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @dimakuv and @woju)

___
*[meson.build, line 16 at r1](https://reviewable.io/reviews/gramineproject/gramine/409#-Mw0aS-M0VMM6kvp4Srg:-Mw22q-SDOqvYKE30tjV:b-d6sehw) ([raw file](https://github.com/gramineproject/gramine/blob/6c82dc4a6063857d7f94b4cfe95c5b34d3fcf77f/meson.build#L16)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Yeah, I'll fix this `env.WERROR`. Looks like Make remnants.

> Do you think people will really use it if you just write about it in README?

The same goes for everything else, like running tests before submitting a PR. We ask for this and then enforce in CI. People often don't do this, but then we can easily see in CI that they didn't. And IMO working by-default is more important than having this small improvement for people who don't read CONTRIBUTING.rst.
</blockquote></details>

Ok, check now.

@pwmarcz 1. and 3. should be fixed now, now sure about 2. (ping @woju).

Also, should I:
a) drop this explicit `'werror=false'`
b) leave as is
c) leave it and add an explicit comment that this is on purpose

Any preferences?

**Review by woju — CHANGES_REQUESTED** [review]
> Reviewed 6 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/409)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @eli-schwartz and @mkow)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/409#-Mw28T80ASgiv7Cvdasu:-Mw28T80ASgiv7Cvdasv:bc25brk) (no related file):*
Just to be sure: you're purposefully omitting this in `cloud-deployment.rst`?

---
___
*[meson.build, line 16 at r1](https://reviewable.io/reviews/gramineproject/gramine/409#-Mw0aS-M0VMM6kvp4Srg:-Mw282tV9T31uRE_dIka:b-sadh07) ([raw file](https://github.com/gramineproject/gramine/blob/6c82dc4a6063857d7f94b4cfe95c5b34d3fcf77f/meson.build#L16)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Ok, check now.

@pwmarcz 1. and 3. should be fixed now, now sure about 2. (ping @woju).

Also, should I:
a) drop this explicit `'werror=false'`
b) leave as is
c) leave it and add an explicit comment that this is on purpose

Any preferences?
</blockquote></details>

I'd defer to @eli-schwartz, please just remove this line (a).
___
*[Documentation/devel/building.rst, line 162 at r2](https://reviewable.io/reviews/gramineproject/gramine/409#-Mw28r7GDOnIr7APTZCz:-Mw28r7H64qyZDwZjkoa:bex5mzx) ([raw file](https://github.com/gramineproject/gramine/blob/7a46c6ee82d833257f05a7cb4813d6f79fbd18e8/Documentation/devel/building.rst#L162)):*
> ```reStructuredText
> 
>    If you plan to contribute changes to Gramine, then you should always build it
>    with ``--werror`` added to the invocation above.
> ```

IDK if this note shouldn't be added below, near `--werror`. Not blocking.

**Review by mkow — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/409)* status: 6 of 7 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @dimakuv, @eli-schwartz, and @woju)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/409#-Mw28T80ASgiv7Cvdasu:-Mw2CN9MCT9bMdw9Za0c:biibotr) (no related file):*
<details><summary><i>Previously, woju (Wojtek Porczyk) wrote…</i></summary><blockquote>

Just to be sure: you're purposefully omitting this in `cloud-deployment.rst`?
</blockquote></details>

Yes, I skipped it because this guide is intended for users, not Gramine developers.

---
___
*[meson.build, line 16 at r1](https://reviewable.io/reviews/gramineproject/gramine/409#-Mw0aS-M0VMM6kvp4Srg:-Mw2CYvf2pVSOpdxWdih:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/6c82dc4a6063857d7f94b4cfe95c5b34d3fcf77f/meson.build#L16)):*
<details><summary><i>Previously, woju (Wojtek Porczyk) wrote…</i></summary><blockquote>

I'd defer to @eli-schwartz, please just remove this line (a).
</blockquote></details>

Done.
___
*[Documentation/devel/building.rst, line 162 at r2](https://reviewable.io/reviews/gramineproject/gramine/409#-Mw28r7GDOnIr7APTZCz:-Mw2CnGFDYDiniomR4Q1:bo2bccf) ([raw file](https://github.com/gramineproject/gramine/blob/7a46c6ee82d833257f05a7cb4813d6f79fbd18e8/Documentation/devel/building.rst#L162)):*
<details><summary><i>Previously, woju (Wojtek Porczyk) wrote…</i></summary><blockquote>

IDK if this note shouldn't be added below, near `--werror`. Not blocking.
</blockquote></details>

I prefer it here to be closer to the command line invocation it conditionally modifies. Less chance that people will miss this ;)

**Review by woju — APPROVED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/409)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @eli-schwartz)

**Review by dimakuv — APPROVED** [review]
> Reviewed 6 of 6 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/409)* status: all files reviewed, all discussions resolved, "fixup! " found in commit messages' one-liners

**Review by dimakuv — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/409)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required)

**Review by woju — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/409)* status: :shipit: complete! all files reviewed, all discussions resolved

**Pull request merged**
